### PR TITLE
refactor: introduce unions via discriminated literal fields

### PR DESCRIPTION
**Pull request by fubuloubu**
### What I did
Introduce more type safety into this library

### How I did it
[Discriminated Unions](https://pydantic-docs.helpmanual.io/usage/types/#discriminated-unions-aka-tagged-unions)

### How to verify it
Tests pass... need to check with `ape`

### Checklist
- [x] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [x] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

**Comment by fubuloubu** [comment]
tested locally with `ape` and works!

File: ethpm_types/abi.py [hunk 1]
```diff
@@ -18,69 +22,150 @@ def canonical_type(self) -> str:
...
+    internalType: Optional[str] = None
+
+    @property
+    def canonical_type(self) -> str:
+        if isinstance(self.type, str):
+            return self.type
+
+        else:
+            # Recursively discover the canonical type
+            return self.type.canonical_type
```
Comment (evanblank3) [inline ethpm_types/abi.py:46]: Would this make sense? And similar treatment for the ABI base class?

```suggestion
class ABIType(BaseModel):
    name: Optional[str] = None  # NOTE: Tuples don't have names by default
    type: Union[str, "ABIType"]
    internalType: Optional[str] = None

    @property
    def canonical_type(self) -> str:
        if isinstance(self.type, str):
            return self.type
        else:
            # Recursively discover the canonical type
            return self.type.canonical_type

    @property
    def signature(self) -> str:
        if self.name:
            return f"{self.canonical_type} {self.name}"
        else:
            return self.canonical_type


class EventABIType(ABIType):
    indexed: bool = False  # Only event ABI types should have this field

    @property
    def signature(self) -> str:
        sig = self.canonical_type
        # For events (handles both None and False conditions)
        if self.indexed:
            sig += " indexed"
        if self.name:
            sig += f" {self.name}"
        return sig

```

Comment (fubuloubu) [reply ethpm_types/abi.py:46]: Fair, I think this could work yeah

**Review by evanblank3 — APPROVED** [review]
:shipit:

**Comment by fubuloubu** [comment]
Works with `ape` main branch and `simple-nft`, also a personal repo of mine. Considering this solid.

**Pull request merged**